File: go/sorting/bubble_sort.go
package sorting

type BubbleSort struct {
}

func (b BubbleSort) Compare(x int, y int, sortType SortType) bool {
	if sortType == "asc" {
		return x > y
	} else {
		return x < y
	}
}

func (b BubbleSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)

	for i := 0; i < length; i++ {
		for j := 0; j < length-i-1; j++ {
			if b.Compare(arr[j], arr[j+1], sortType) {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}

	return arr
}

File: dart/lib/sorting/bubble_sort.dart
import 'package:algorithm/sorting/sort.dart';

class BubbleSort extends Sort {
  @override
  List<int> sort(List<int> arr, {SortType type = SortType.asc}) {
    final length = arr.length;

    for (var i = 0; i < length; i++) {
      for (var j = 0; j < length - i - 1; j++) {
        if (compare(arr[j], arr[j + 1], type)) {
          arr.swap(j, j + 1);
        }
      }
    }

    return arr;
  }
}
